In [1]:
import sys
from pathlib import Path

from dotenv import load_dotenv
load_dotenv()

True

In [2]:
# LangChain 관련 라이브러리
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings, ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_huggingface import HuggingFaceEmbeddings

In [3]:
# RAGAS 관련 라이브러리
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper
from ragas.testset import TestsetGenerator
from ragas import EvaluationDataset, evaluate
from ragas.metrics import LLMContextRecall, Faithfulness, FactualCorrectness

# 벡터 저장소 생성/저장

In [4]:
embedding_model = HuggingFaceEmbeddings(model_name="BAAI/bge-m3")

# 기존 벡터 저장소 로드
vector_store = Chroma(
    embedding_function=embedding_model,
    persist_directory="./utils/chroma_db",
)

# 결과 확인
print(f"저장된 Document 개수: {len(vector_store.get()['ids'])}")

저장된 Document 개수: 249


In [6]:
from ragas.testset.persona import Persona

# 페르소나 정의
personas = [
    Persona(
        name="backend_team",
        role_description="CTO의 문서에 접근가능한 평사원. 한국어 사용자",
    ),
    Persona(
        name="frontend_team",
        role_description="프론트앤드 팀의 문서에 접근가능한 평사원. 한국어 사용자",
    ),
    Persona(
        name="data_ai_team",
        role_description="AI 팀의 문서에 접근가능한 평사원. 한국어 사용자",
    ),
]

In [7]:
# LLM과 임베딩 모델 초기화
generator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o-mini"))
generator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings(model="text-embedding-3-small"))

C:\Users\playdata\AppData\Local\Temp\ipykernel_25104\2522793028.py:3: DeprecationWarning: LangchainEmbeddingsWrapper is deprecated and will be removed in a future version. Use the modern embedding providers instead: embedding_factory('openai', model='text-embedding-3-small', client=openai_client) or from ragas.embeddings import OpenAIEmbeddings, GoogleEmbeddings, HuggingFaceEmbeddings
  generator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings(model="text-embedding-3-small"))


In [8]:
# TestsetGenerator 생성
generator = TestsetGenerator(
    llm=generator_llm,
    embedding_model=generator_embeddings,
    persona_list=personas
)

In [9]:
# Load and Adapt Queries
# Here we load the required query types and adapt them to the target language.
from ragas.testset.synthesizers.single_hop.specific import (
    SingleHopSpecificQuerySynthesizer,
)

distribution = [
    (SingleHopSpecificQuerySynthesizer(llm=generator_llm), 1.0),
]

for query, _ in distribution:
    prompts = await query.adapt_prompts("korean", llm=generator_llm)
    query.set_prompts(**prompts)

In [14]:
from ragas.testset.transforms.extractors.llm_based import NERExtractor
from ragas.testset.transforms.splitters import HeadlineSplitter

transforms = [NERExtractor()]

In [15]:
from langchain.schema import Document

# Chroma에서 문서와 메타데이터 불러오기
raw = vector_store.get(include=["documents", "metadatas"])

# Document 리스트로 변환
docs = [
    Document(page_content=doc, metadata=meta)
    for doc, meta in zip(raw["documents"], raw["metadatas"])
]

In [16]:
# generate
dataset = generator.generate_with_langchain_docs(
    docs[:20],
    testset_size=20,
    transforms=transforms,
    query_distribution=distribution,
)

Applying NERExtractor:   0%|          | 0/20 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/1 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/12 [00:00<?, ?it/s]

In [17]:
# 데이터셋을 DataFrame으로 변환
import pandas as pd

# 컬럼너비 제한 없음
pd.set_option('display.max_colwidth', None)


dataset_df = dataset.to_pandas()
display(dataset_df)

user_input  \
0                                          크랙(Crack) 앱의 주요 기능은 무엇인가요?   
1                                                RabbitMQ의 역할은 무엇인가요?   
2                                           프론트엔드 팀에서 JWT의 역할은 무엇인가요?   
3                         코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?   
4                                                         MFA는 뭐하는거야?   
5                                           JWT는 API 접근 시 어떤 역할을 합니까?   
6                       코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서 에러 핸들링의 중요성은 무엇으로 설명되고 있습니까?   
8                                                           v1.0 뭐에요?   
9             코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목들은 무엇인가요?   
10                                                      챗봇 대화는 뭐 저장해?   
11                                      2025-08-01에 기록된 변경 사항은 무엇입니까?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       reference_contexts  \
0                                              [<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->\n# 서비스 아키텍처 문서\n분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29\n\n## 1. 개요\n코드노바의 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 이 문서는 백엔드 시스템의 구성 요소와 상호작용을 설명합니다.\n\n## 2. 아키텍처 구성 요소\n\n### 2.1. API 서버\n- **역할**: 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환합니다.\n- **기술 스택**: Node.js, Express.js\n- **검증 포인트**:\n  - API 엔드포인트가 올바르게 작동하는지 확인\n  - 요청 처리 속도 및 오류율 모니터링]   
1                                                                                            [### 2.2. 데이터베이스\n- **역할**: 사용자 데이터, 콘텐츠, 로그 등을 저장합니다.\n- **기술 스택**: PostgreSQL\n- **검증 포인트**:\n  - 데이터베이스 연결 상태 확인\n  - 쿼리 성능 분석 및 최적화\n\n### 2.3. 캐시 서버\n- **역할**: 데이터베이스 부하를 줄이고 응답 속도를 높이기 위해 자주 조회되는 데이터를 캐싱합니다.\n- **기술 스택**: Redis\n- **검증 포인트**:\n  - 캐시 적중률 모니터링\n  - 캐시 데이터의 일관성 확인\n\n### 2.4. 메시지 큐\n- **역할**: 비동기 작업 처리를 위해 시스템 간 메시지를 전달합니다.\n- **기술 스택**: RabbitMQ\n- **검증 포인트**:\n  - 메시지 전송 성공률 확인\n  - 큐의 길이 및 처리 속도 모니터링]   
2                [### 2.5. 파일 저장소\n- **역할**: 이미지 및 기타 미디어 파일을 저장합니다.\n- **기술 스택**: AWS S3\n- **검증 포인트**:\n  - 파일 업로드 및 다운로드 기능 확인\n  - 저장소 사용량 모니터링\n\n## 3. 서비스 흐름\n\n1. **사용자 요청**: 클라이언트에서 API 서버에 요청을 보냅니다.\n2. **API 처리**: API 서버는 요청을 처리하고 필요한 경우 데이터베이스 또는 캐시 서버에 접근합니다.\n3. **비동기 작업**: 일부 요청은 메시지 큐에 전송되어 비동기로 처리됩니다.\n4. **응답 반환**: API 서버는 클라이언트에 응답을 반환합니다.\n\n## 4. 보안\n- **인증**: JWT(JSON Web Token)를 사용하여 사용자 인증을 처리합니다.\n- **데이터 암호화**: 전송 중 데이터 암호화를 위해 HTTPS를 사용합니다.\n- **검증 포인트**:\n  - 인증 토큰의 유효성 검사\n  - 보안 로그 모니터링]   
3                                               [## 5. 모니터링 및 로깅\n- **모니터링 도구**: Prometheus 및 Grafana를 사용하여 시스템 성능 및 상태를 모니터링합니다.\n- **로깅**: ELK 스택(Elasticsearch, Logstash, Kibana)을 사용하여 로그를 수집하고 분석합니다.\n- **검증 포인트**:\n  - 주요 메트릭스 확인\n  - 시스템 오류 및 경고 모니터링\n\n## 6. 결론\n코드노바의 백엔드 서비스 아키텍처는 확장 가능하고 안정적인 서비스를 제공하기 위해 설계되었습니다. 각 구성 요소는 독립적으로 운영되며, 서로 긴밀하게 연결되어 있습니다. 지속적인 모니터링과 최적화를 통해 서비스의 품질을 유지하고 향상시킬 수 있습니다.\n\n---\n\n다음 개정 제안: 서비스 아키텍처의 각 구성 요소에 대한 세부적인 기술 스택과 버전 정보를 추가하는 것이 좋습니다. 또한, 보안 관련 사항을 더 구체적으로 설명할 필요가 있습니다.]   
4   [<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->\n# 보안/인증 가이드\n분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29\n\n## 1. 보안 기본 원칙\n\n### 1.1 최소 권한 원칙\n- 사용자는 자신의 업무에 필요한 최소한의 권한만 부여받아야 합니다.\n- 정기적으로 권한을 검토하고 불필요한 권한은 즉시 회수합니다.\n\n### 1.2 데이터 암호화\n- 모든 민감 데이터는 저장 시 암호화하여 보호합니다.\n- 전송되는 데이터는 SSL/TLS 프로토콜을 사용하여 암호화합니다.\n\n### 1.3 정기적인 보안 점검\n- 보안 점검을 정기적으로 실시하여 취약점을 발견하고 수정합니다.\n- 외부 보안 전문가에 의한 감사도 고려합니다.\n\n## 2. 사용자 인증\n\n### 2.1 인증 방법\n- 기본 인증 방식으로는 이메일과 비밀번호 조합을 사용합니다.\

In [10]:
import pandas as pd

dataset_df = pd.read_csv('ragas_dataset1.csv')

display(dataset_df)

user_input  \
0                          크랙(Crack) 앱의 주요 기능은 무엇인가요?   
1                                RabbitMQ의 역할은 무엇인가요?   
2                           프론트엔드 팀에서 JWT의 역할은 무엇인가요?   
3         코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?   
4                                         MFA는 뭐하는거야?   
5                           JWT는 API 접근 시 어떤 역할을 합니까?   
6       코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서 에러 핸들링의 중...   
8                                           v1.0 뭐에요?   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10                                      챗봇 대화는 뭐 저장해?   
11                      2025-08-01에 기록된 변경 사항은 무엇입니까?   

                                   reference_contexts  \
0   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
1   ['### 2.2. 데이터베이스\n- **역할**: 사용자 데이터, 콘텐츠, 로그 ...   
2   ['### 2.5. 파일 저장소\n- **역할**: 이미지 및 기타 미디어 파일을 ...   
3   ['## 5. 모니터링 및 로깅\n- **모니터링 도구**: Prometheus 및...   
4   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
5   ['### 2.2 비밀번호 관리\n- 비밀번호는 최소 8자 이상, 대문자, 소문자,...   
6   ['## 4. 로그 관리\n\n### 4.1 로그 기록\n- 모든 인증 시도 및 A...   
7   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
8   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
9   ['## 체크리스트\n\n- [ ] 코드 리뷰 완료\n- [ ] 모든 테스트 통과\...   
10  ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
11  ["### 3.1 변경 이력 기록 양식\n\n- **변경 날짜**: YYYY-MM-...   

                                            reference  \
0   크랙(Crack)은 AI 페르소나 챗봇 앱으로, 코드노바의 서비스 아키텍처에서 생성...   
1   RabbitMQ는 비동기 작업 처리를 위해 시스템 간 메시지를 전달하는 메시지 큐입니다.   
2          JWT(JSON Web Token)는 사용자 인증을 처리하는 데 사용됩니다.   
3   코드노바의 백엔드 서비스 아키텍처는 확장 가능하고 안정적인 서비스를 제공하기 위해 ...   
4               MFA는 추가적인 보안 강화를 위해 도입되는 다중 인증 방법입니다.   
5   JWT(JSON Web Token)는 API 접근 시 인증을 위해 사용되며, 토큰은...   
6   이 가이드는 코드노바의 백엔드 개발팀이 보안 및 인증을 효과적으로 관리하기 위한 기...   
7   코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링이 안정적인 시스템 운영을 위해 필수...   
8   v1.0은 코드노바 백엔드팀의 배포 및 운영 가이드의 버전 번호로, 이 문서는 안정...   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10                          챗봇 대화는 챗봇과의 대화 기록을 저장합니다.   
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...   

                         synthesizer_name  
0   single_hop_specific_query_synthesizer  
1   single_hop_specific_query_synthesizer  
2   single_hop_specific_query_synthesizer  
3   single_hop_specific_query_synthesizer  
4   single_hop_specific_query_synthesizer  
5   single_hop_specific_query_synthesizer  
6   single_hop_specific_query_synthesizer  
7   single_hop_specific_query_synthesizer  
8   single_hop_specific_query_synthesizer  
9   single_hop_specific_query_synthesizer  
10  single_hop_specific_query_synthesizer  
11  single_hop_specific_query_synthesizer

## 02.RAG 체인 구성

검색기와 생성 모델을 결합한 RAG 체인을 구성한다.

In [18]:
from dotenv import load_dotenv

load_dotenv()

from pathlib import Path
import sys

root = Path.cwd()
while not (root / "utils2").exists() and root != root.parent:
    root = root.parent
sys.path.insert(0, str(root))

from utils2.rag import chat

response = chat({"history": [
    {"role": "user", "content": "코드노바의 API 서버 기술스택알려줘"}], "tone":"informal", "permission":"cto"})

INFO:utils2.service:Chat Request - Permission: cto, Tone: informal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "코드노바의 API 서버 기술스택알려줘"}}']
INFO:utils2.service:cto search keyword: 코드노바의 API 서버 기술스택알려줘


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 서비스 아키텍처 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 개요
코드노바의 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 이 문서는 백엔드 시스템의 구성 요소와 상호작용을 설명합니다.

## 2. 아키텍처 구성 요소

### 2.1. API 서버
- **역할**: 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환합니다.
- **기술 스택**: Node.js, Express.js
- **검증 포인트**:
  - API 엔드포인트가 올바르게 작동하는지 확인
  - 요청 처리 속도 및 오류율 모니터링 [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# API 정책 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 목적
이 문서는 코드노바의 백엔드팀에서 API를 설계하고 운영하는 데 필요한 정책과 절차를 정의합니다. API의 일관성과 안정성을 보장하고, 외부 개발자 및 내부 팀 간의 원활한 협업을 촉진하기 위한 것입니다. [[ref2]]
<!-- 회사: 코드노바 | 대상: 사원(프론트엔드) | 작성일: 2025-08-29 -->
# 프론트엔드 아키텍처 문서
분류: frontend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

---

## 1. 개요
코드노바의 프론트엔드 아키텍처는 사용자 경험을 극대화하고, 유지보수성을 높이며, 확장성을 고려하여 설계되었습니다. 본 문서는 프론트엔드 개발자가 이해하고 실행할 수 있도록 아키텍처의 구성 요소와 원칙을 설명합니다.

## 2. 아키텍처 구성 

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:코드노바의 API 서버 기술스택은 Node.js와 Express.js로 구성되어 있습니다. 이 API 서버는 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환하는 역할을 합니다.


In [19]:
print(response['response'])

코드노바의 API 서버 기술스택은 Node.js와 Express.js로 구성되어 있습니다. 이 API 서버는 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환하는 역할을 합니다.


In [3]:
print(response['result'])
type(response['result'])

검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 서비스 아키텍처 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 개요
코드노바의 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 이 문서는 백엔드 시스템의 구성 요소와 상호작용을 설명합니다.

## 2. 아키텍처 구성 요소

### 2.1. API 서버
- **역할**: 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환합니다.
- **기술 스택**: Node.js, Express.js
- **검증 포인트**:
  - API 엔드포인트가 올바르게 작동하는지 확인
  - 요청 처리 속도 및 오류율 모니터링 [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# API 정책 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 목적
이 문서는 코드노바의 백엔드팀에서 API를 설계하고 운영하는 데 필요한 정책과 절차를 정의합니다. API의 일관성과 안정성을 보장하고, 외부 개발자 및 내부 팀 간의 원활한 협업을 촉진하기 위한 것입니다. [[ref2]]
<!-- 회사: 코드노바 | 대상: 사원(프론트엔드) | 작성일: 2025-08-29 -->
# 프론트엔드 아키텍처 문서
분류: frontend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

---

## 1. 개요
코드노바의 프론트엔드 아키텍처는 사용자 경험을 극대화하고, 유지보수성을 높이며, 확장성을 고려하여 설계되었습니다. 본 문서는 프론트엔드 개발자가 이해하고 실행할 수 있도록 아키텍처의 구성 요소와 원칙을 설명합니다.

## 2. 아키텍처 구성 

str

## 03.RAGAS기반 평가

RAGAS는 RAG(Retrieval-Augmented Generation) 시스템을 평가하는 자동화된 참조 없는(reference-free) 평가 프레임워크입니다. 즉, 사람이 만든 정답 데이터(ground truth)가 없이도 RAG 시스템의 검색과 생성 단계를 자동으로 평가할 수 있도록 설계된 도구입니다. RAGAS는 LLM을 활용해 평가를 수행하며, 검색된 컨텍스트와 생성된 답변의 충실도, 관련성, 정밀도, 재현율 등 다양한 측면을 측정합니다.

### RAGAS 주요 평가 지표
- **Faithfulness(충실도)**: 생성된 답변이 주어진 컨텍스트 정보에 얼마나 충실한지를 평가합니다. 답변 내용이 컨텍스트에서 실제로 뒷받침되는지 보는 지표입니다.
- **Answer Relevancy(답변 관련성)**: 답변이 원 질문과 얼마나 관련성이 높은지를 측정합니다.
- **Context Precision(컨텍스트 정밀도)**: 검색된 컨텍스트 문서가 질문에 적절한 정보인지, 관련된 문서가 상위에 있는지를 평가합니다.
- **Context Recall(컨텍스트 재현율)**: 답변을 생성하는 데 필요한 컨텍스트를 얼마나 잘 검색했는지 평가합니다.


In [5]:
# 평가용 데이터 로드
import pandas as pd

# 같은 폴더(현재 작업 디렉토리)에 저장했다면
dataset_df = pd.read_csv("./ragas_dataset1.csv")
print(dataset_df.shape)

eval_dataset = dataset_df[['user_input', 'reference_contexts', 'reference']]
display(eval_dataset)

(12, 4)


user_input  \
0                          크랙(Crack) 앱의 주요 기능은 무엇인가요?   
1                                RabbitMQ의 역할은 무엇인가요?   
2                           프론트엔드 팀에서 JWT의 역할은 무엇인가요?   
3         코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?   
4                                         MFA는 뭐하는거야?   
5                           JWT는 API 접근 시 어떤 역할을 합니까?   
6       코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서 에러 핸들링의 중...   
8                                           v1.0 뭐에요?   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10                                      챗봇 대화는 뭐 저장해?   
11                      2025-08-01에 기록된 변경 사항은 무엇입니까?   

                                   reference_contexts  \
0   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
1   ['### 2.2. 데이터베이스\n- **역할**: 사용자 데이터, 콘텐츠, 로그 ...   
2   ['### 2.5. 파일 저장소\n- **역할**: 이미지 및 기타 미디어 파일을 ...   
3   ['## 5. 모니터링 및 로깅\n- **모니터링 도구**: Prometheus 및...   
4   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
5   ['### 2.2 비밀번호 관리\n- 비밀번호는 최소 8자 이상, 대문자, 소문자,...   
6   ['## 4. 로그 관리\n\n### 4.1 로그 기록\n- 모든 인증 시도 및 A...   
7   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
8   ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
9   ['## 체크리스트\n\n- [ ] 코드 리뷰 완료\n- [ ] 모든 테스트 통과\...   
10  ['<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-2...   
11  ["### 3.1 변경 이력 기록 양식\n\n- **변경 날짜**: YYYY-MM-...   

                                            reference  
0   크랙(Crack)은 AI 페르소나 챗봇 앱으로, 코드노바의 서비스 아키텍처에서 생성...  
1   RabbitMQ는 비동기 작업 처리를 위해 시스템 간 메시지를 전달하는 메시지 큐입니다.  
2          JWT(JSON Web Token)는 사용자 인증을 처리하는 데 사용됩니다.  
3   코드노바의 백엔드 서비스 아키텍처는 확장 가능하고 안정적인 서비스를 제공하기 위해 ...  
4               MFA는 추가적인 보안 강화를 위해 도입되는 다중 인증 방법입니다.  
5   JWT(JSON Web Token)는 API 접근 시 인증을 위해 사용되며, 토큰은...  
6   이 가이드는 코드노바의 백엔드 개발팀이 보안 및 인증을 효과적으로 관리하기 위한 기...  
7   코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링이 안정적인 시스템 운영을 위해 필수...  
8   v1.0은 코드노바 백엔드팀의 배포 및 운영 가이드의 버전 번호로, 이 문서는 안정...  
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...  
10                          챗봇 대화는 챗봇과의 대화 기록을 저장합니다.  
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...

In [9]:
# 평가용 데이터셋 생성
evaluated_dataset = []

# 각 행에 대해 RAG 체인을 호출하여 결과를 저장
for row in eval_dataset.itertuples():
    query = row.user_input  # 사용자 입력
    response = chat({"history": [
    {"role": "user", "content": query}], "tone":"formal", "permission":"cto"})
    retrieved_contexts = response['result']  # 실제 검색된 문서

    reference = row.reference  # 정답
    reference_contexts = row.reference_contexts  # 정답 참조 컨텍스트


    evaluated_dataset.append(
        {
            "user_input": query,
            "retrieved_contexts": [retrieved_contexts],
            "response": response['response'],
            "reference": reference,
            # "reference_contexts": reference_contexts,
        }
    )

# RAGAS 평가 데이터셋 생성
ragas_evaluated_dataset = EvaluationDataset.from_list(evaluated_dataset)

# 데이터 저장
ragas_evaluated_dataset.to_pandas().to_csv('ragas_evaluated_dataset.csv', index=False)

INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "크랙(Crack) 앱의 주요 기능은 무엇인가요?"}}']
INFO:utils2.service:cto search keyword: 크랙(Crack) 앱의 주요 기능은 무엇인가요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 서비스 아키텍처 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 개요
코드노바의 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 이 문서는 백엔드 시스템의 구성 요소와 상호작용을 설명합니다.

## 2. 아키텍처 구성 요소

### 2.1. API 서버
- **역할**: 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환합니다.
- **기술 스택**: Node.js, Express.js
- **검증 포인트**:
  - API 엔드포인트가 올바르게 작동하는지 확인
  - 요청 처리 속도 및 오류율 모니터링 [[ref1]]
## 6. 의존성 관리
- **패키지 업데이트**: 사용하는 라이브러리와 패키지는 정기적으로 업데이트하여 최신 보안 패치를 적용합니다.
- **취약점 스캐닝**: 의존성에 대한 취약점 스캐닝 도구를 사용하여 보안 취약점을 사전 예방적으로 점검합니다.

## 7. 오류 처리
- **사용자 친화적인 오류 메시지**: 오류 발생 시, 사용자에게는 일반적인 오류 메시지만 표시하고, 내부 정보는 노출하지 않도록 합니다.
- **로깅 및 모니터링**: 오류 발생 시, 로그를 기록하고, 이를 통해 보안 사고를 추적할 수 있도록 합니다.

## 8. 배포 보안
- **배포 환경 설정**: 배포 환경은 개발 환경과 분리하여 설정하고, 필요한 경우 방화벽을 설정합니다.
- **CI/CD 파이프라인 보안**: CI/CD 파이프라인의 접근 권한을 제한하고, 보안 점검을 자동화하여 배포 전 검증합니다. [[ref2]]
### 2.2. 데이터베이스
- **역할**: 사용자 데이터, 콘텐츠, 로그 등을 저장합니다.
- **기술 스택**: Po

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:크랙(Crack) 앱의 주요 기능은 생성형 AI를 활용한 글쓰기, 이미지 생성, 요약 플랫폼을 지원하는 것입니다. 또한, AI 페르소나 챗봇 앱으로서 대화형 광고 제작 및 보상 플랫폼인 Wrtn Ads와도 통합되어 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "RabbitMQ의 역할은 무엇인가요?"}}']
INFO:utils2.service:cto search keyword: RabbitMQ의 역할은 무엇인가요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
### 2.2. 데이터베이스
- **역할**: 사용자 데이터, 콘텐츠, 로그 등을 저장합니다.
- **기술 스택**: PostgreSQL
- **검증 포인트**:
  - 데이터베이스 연결 상태 확인
  - 쿼리 성능 분석 및 최적화

### 2.3. 캐시 서버
- **역할**: 데이터베이스 부하를 줄이고 응답 속도를 높이기 위해 자주 조회되는 데이터를 캐싱합니다.
- **기술 스택**: Redis
- **검증 포인트**:
  - 캐시 적중률 모니터링
  - 캐시 데이터의 일관성 확인

### 2.4. 메시지 큐
- **역할**: 비동기 작업 처리를 위해 시스템 간 메시지를 전달합니다.
- **기술 스택**: RabbitMQ
- **검증 포인트**:
  - 메시지 전송 성공률 확인
  - 큐의 길이 및 처리 속도 모니터링 [[ref1]]
---

## 2. 팀원별 주요 작업
### 김철수 (데이터 품질 관리)
- 데이터 오류 검토 및 수정
  - 의존성: 데이터 수집팀의 피드백 필요
  - 리스크: 데이터 수집 지연 시 작업 지연 가능성

- 품질 개선 방안 제안서 작성
  - 의존성: 데이터 분석팀의 협조 필요
  - 리스크: 제안서 승인 지연 시 적용 시점 늦어질 수 있음

### 이영희 (AI 모델 개발)
- 모델 성능 평가 및 개선
  - 의존성: 최신 데이터셋 확보 필요
  - 리스크: 데이터셋 품질 저하 시 성능 저하 우려

- 코드 리뷰 및 테스트
  - 의존성: 팀원들과의 협업 필요
  - 리스크: 코드 충돌 발생 가능성

### 박민수 (데이터 엔지니어링)
- 신규 파이프라인 설계 및 구현
  - 의존성: 기존 시스템과의 호환성 검토 필요
  - 리스크: 기술적 문제 발생 시 일정 지연 가능성

- 문서화 작업
  - 의존성: 관련 팀의 피드백 필요
  - 리스크: 문서화 지연 시 지식 공유 부족 [[ref2]]
# 모델 개발 & 성능 | 모델 성능 평가 보고서

작성일: 2025-08-29
회사: CodeNova | 대상

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:RabbitMQ는 메시지 큐로, 비동기 작업 처리를 위해 시스템 간 메시지를 전달하는 역할을 합니다. 주요 기능은 메시지 전송 성공률 확인과 큐의 길이 및 처리 속도 모니터링입니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "프론트엔드 팀에서 JWT의 역할은 무엇인가요?"}}']
INFO:utils2.service:cto search keyword: 프론트엔드 팀에서 JWT의 역할은 무엇인가요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 6. 개정 이력
- v1.0 — 오늘

--- 

이 문서는 프론트엔드팀의 근무 태도 및 기술 기여도를 종합적으로 평가하여 CTO의 의사결정에 필요한 정보를 제공합니다. 각 항목에 대한 체크리스트를 통해 실행 계획을 수립하고, 리스크 관리에 대한 전략을 마련하는 데 기여할 수 있습니다. [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(프론트엔드) | 작성일: 2025-08-29 -->
# 협업 프로세스 문서
분류: frontend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 팀 구성원 역할 정의
- **프론트엔드 개발자**: UI/UX 구현, 코드 리뷰 및 버그 수정
- **디자이너**: UI 디자인 및 사용자 경험 설계
- **프로덕트 매니저**: 요구사항 수집 및 프로젝트 관리
- **QA 엔지니어**: 품질 보증 및 테스트

## 2. 협업 도구
- **커뮤니케이션**: 슬랙(Slack) 또는 팀즈(Teams) 활용
- **프로젝트 관리**: 지라(Jira) 또는 트렐로(Trello) 사용
- **버전 관리**: 깃(Git) 및 깃허브(GitHub) 활용

## 3. 협업 프로세스 단계 [[ref2]]
# 팀 성과 / 내부 평가 | 내부 평가 피드백 문서_프론트엔드팀 근무 태도 및 기술 기여도 평가

작성일: 2025-08-29
회사: CodeNova | 대상: CTO

---
# 내부 평가 피드백 문서_프론트엔드팀 근무 태도 및 기술 기여도 평가
(분류: 팀 성과 / 내부 평가) | 회사: CodeNova | 버전: v1.0 | 작성일: 2025-08-29

---

## 1. 평가 목적 및 기준
- **목적**: 프론트엔드팀의 근무 태도 및 기술 기여도를 평가하여 팀 성과를 분석하고 향후 개선 방향을 제시.
- **기준**:
  - 업무 수행 능력
  - 팀워크 및 커뮤니케이션
  - 기술적 기여도 (프로젝트 기여 및 혁신)
  - 문제 해결 능력

## 2. 평가 대상 및 

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:검색 결과에 프론트엔드 팀에서 JWT의 역할에 대한 구체적인 내용은 포함되어 있지 않습니다. 추가로 필요한 정보가 있으시면 말씀해 주시기 바랍니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?"}}']
INFO:utils2.service:cto search keyword: 코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 서비스 아키텍처 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 개요
코드노바의 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 이 문서는 백엔드 시스템의 구성 요소와 상호작용을 설명합니다.

## 2. 아키텍처 구성 요소

### 2.1. API 서버
- **역할**: 클라이언트와의 통신을 담당하며, 요청을 처리하고 적절한 응답을 반환합니다.
- **기술 스택**: Node.js, Express.js
- **검증 포인트**:
  - API 엔드포인트가 올바르게 작동하는지 확인
  - 요청 처리 속도 및 오류율 모니터링 [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 백엔드팀 주간 업무 계획서
분류: meeting | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29
---

## 주간 목표
- 서비스 안정성 향상
- 새로운 기능 개발 및 테스트
- 성능 최적화 및 버그 수정

## 주요 업무
1. **기능 개발**
   - [ ] 신규 기능 개발을 위한 요구사항 정리
   - [ ] API 설계 및 문서화
   - [ ] 기능 구현 및 코드 리뷰 진행
   - [ ] 단위 테스트 작성 및 통과 확인

2. **버그 수정**
   - [ ] 기존 버그 리스트 검토
   - [ ] 우선순위에 따른 버그 수정 작업 진행
   - [ ] 수정된 버그에 대한 테스트 및 검증
   - [ ] 버그 수정 결과 문서화 [[ref2]]
<!-- 회사: 코드노바 | 대상: 사원(프론트엔드) | 작성일: 2025-08-29 -->
# 프론트엔드 아키텍처 문서
분류: frontend | 회

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:코드노바의 백엔드 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI 페르소나 챗봇 앱인 크랙(Crack), 그리고 대화형 광고 제작·보상 플랫폼인 Wrtn Ads를 지원하도록 설계되었습니다. 주요 구성 요소로는 API 서버가 있으며, 이 서버는 클라이언트와의 통신을 담당하고 요청을 처리하여 적절한 응답을 반환합니다. API 서버는 Node.js와 Express.js 기술 스택을 사용합니다. 

이 아키텍처의 목적은 다양한 AI 기반 서비스를 안정적으로 지원하고, 효율적인 요청 처리 및 응답을 통해 시스템의 성능과 안정성을 확보하는 데 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "MFA는 뭐하는거야?"}}']
INFO:utils2.service:ct

name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 5. 리스크 한계 및 보완책
- **리스크 한계**: 키 유출 및 무단 접근의 가능성.
- **보완책**:
  - 다중 인증(MFA) 도입: 키 접근 시 추가 인증 절차 요구.
  - 정기적인 보안 감사 실시: 외부 전문가에 의한 보안 점검.

## 6. CTO 보고·승인 포인트
- 키 관리 정책 수립 및 변경 사항에 대한 정기 보고.
- 주요 보안 사건 발생 시 즉각적인 보고 및 대응 방안 승인 요청.

## 7. 개정 이력
- v1.0 — 오늘: 최초 작성.

--- 

이 문서는 CodeNova의 민감 데이터 접근 키 및 암호화 키 관리에 대한 전략적 접근을 제시하며, CTO의 의사결정 및 리스크 관리에 기여할 것입니다. [[ref1]]
- **기능 B 개발**:
  - [ ] 기획서 검토 및 피드백 반영
  - [ ] 데이터베이스 설계 변경 사항 논의
  - [ ] 개발 진행 상황 주간 점검

### 2.3 기술 부채 관리
- **리팩토링**:
  - [ ] 코드 품질 점검 도구 실행
  - [ ] 기술 부채 목록 업데이트
  - [ ] 우선 순위에 따라 리팩토링 계획 수립
  - [ ] 리팩토링 진행 후 성능 테스트 실시

## 3. 회의 일정

- **정기 회의**: 매주 수요일 오후 3시
- **스프린트 회의**: 매주 금요일 오전 10시

## 4. 체크리스트

- [ ] 각 팀원별 주간 업무 목표 공유
- [ ] 진행 중인 업무의 현황 점검
- [ ] 문제 발생 시 해결 방안 논의
- [ ] 다음 주 업무 계획 수립

## 5. 검증 포인트

- 서비스 안정성 지표(응답 시간, 오류율) 확인 및 개선 여부
- 신규 기능 개발 일정 준수 여부
- 기술 부채 관리 진행 상황 및 리팩토링 완료 여부

--- [[ref2]]
## 2. 팀원별 주요 작업
### 이지훈
- **작업**: 모델 성능 개선을 위한 하이퍼파라미터 튜닝
  - **의존성**: 데이터셋 업데이트 완료 필요
  - **리스크**: 성능 개선이 예상보다 

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:MFA는 다중 인증(Multi-Factor Authentication)의 약자로, 시스템이나 데이터에 접근할 때 추가적인 인증 절차를 요구하는 보안 기술입니다. 이는 키 유출 및 무단 접근의 위험을 줄이기 위해 도입되며, 보안 강화를 위한 중요한 보완책 중 하나입니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "JWT는 API 접근 시 어떤 역할을 합니까?"}}']
INFO:utils2.service:cto search keyword: JWT는 API 접근 시 어떤 역할을 합니까?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 3. 인증 및 권한 관리
- **인증 방식**: API는 OAuth 2.0 또는 JWT(JSON Web Token)를 사용하여 인증을 처리합니다. 모든 요청은 적절한 인증 토큰을 포함해야 합니다.
- **권한 부여**: 각 API 엔드포인트에 대한 접근 권한은 역할 기반으로 관리됩니다. 특정 역할에 따라 접근 가능한 자원과 작업이 제한됩니다.

## 4. 오류 처리
- **표준화된 오류 응답**: API는 모든 오류에 대해 일관된 형식의 오류 응답을 반환해야 합니다. 오류 코드와 메시지는 명확하고 이해하기 쉽게 작성되어야 합니다.
- **HTTP 상태 코드 사용**: 각 오류에 대해 적절한 HTTP 상태 코드를 사용하여 클라이언트가 오류의 종류를 쉽게 파악할 수 있도록 합니다. [[ref1]]
### 2.2 비밀번호 관리
- 비밀번호는 최소 8자 이상, 대문자, 소문자, 숫자, 특수문자를 포함해야 합니다.
- 비밀번호는 주기적으로 변경하도록 유도합니다.

### 2.3 세션 관리
- 사용자가 로그인 후 일정 시간 동안 활동이 없으면 자동으로 로그아웃됩니다.
- 세션 ID는 예측 불가능한 값으로 생성하고, 매 요청 시 갱신합니다.

## 3. API 보안

### 3.1 인증 토큰 사용
- API 접근 시 JWT(JSON Web Token) 또는 OAuth2를 사용하여 인증합니다.
- 토큰은 유효 기간을 설정하고, 만료된 토큰은 사용하지 못하도록 합니다.

### 3.2 CORS 설정
- Cross-Origin Resource Sharing(CORS) 정책을 설정하여 허용된 도메인에서만 API 접근을 허용합니다.

### 3.3 요청 검증
- 모든 API 요청에 대해 유효성을 검증합니다.
- SQL 인젝션, XSS 공격 등을 방지하기 위한 필터링을 적용합니다. [[ref2]]
### 2.5. 파일 저장소
- **역할**: 이미지 및 기타 미디어 파일을 저장합니다.
- **기술 스택**: AWS S3
- **검증 포인트**:
  - 파

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:JWT(JSON Web Token)는 API 접근 시 사용자의 인증을 처리하는 데 중요한 역할을 합니다. API 요청 시, 클라이언트는 JWT를 포함하여 서버에 접근합니다. 서버는 이 토큰을 통해 사용자의 신분을 확인하고, 토큰이 유효한지 검증합니다. 또한, JWT는 접근 권한을 관리하는 데 사용되며, 특정 역할에 따라 자원과 작업에 대한 제한을 설정할 수 있습니다. 만료된 토큰은 사용할 수 없도록 설정되어 보안을 강화합니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?"}}']
INFO:utils2.service:cto search keyword: 코드노바의 백엔드 개발팀이 보안 및 인증을

name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 4. 로그 관리

### 4.1 로그 기록
- 모든 인증 시도 및 API 호출에 대한 로그를 기록합니다.
- 로그에는 사용자 ID, IP 주소, 요청 시간, 요청 내용 등을 포함합니다.

### 4.2 로그 분석
- 정기적으로 로그를 분석하여 비정상적인 활동을 탐지합니다.
- 이상 징후 발견 시 즉시 대응합니다.

## 5. 사고 대응

### 5.1 사고 대응 계획
- 보안 사고 발생 시 대응 절차를 문서화합니다.
- 사고 발생 시 즉시 관련 팀에 통보하고, 피해를 최소화하기 위한 조치를 취합니다.

### 5.2 교육 및 훈련
- 모든 팀원은 보안 교육을 정기적으로 이수해야 합니다.
- 보안 사고 대응 훈련을 통해 실제 상황에 대비합니다.

## 6. 결론
이 가이드는 코드노바의 백엔드 개발팀이 보안 및 인증을 효과적으로 관리하기 위한 기본 지침입니다. 모든 팀원은 이 가이드를 숙지하고 준수하여 안전한 시스템 운영에 기여해야 합니다.

--- [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 보안/인증 가이드
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 보안 기본 원칙

### 1.1 최소 권한 원칙
- 사용자는 자신의 업무에 필요한 최소한의 권한만 부여받아야 합니다.
- 정기적으로 권한을 검토하고 불필요한 권한은 즉시 회수합니다.

### 1.2 데이터 암호화
- 모든 민감 데이터는 저장 시 암호화하여 보호합니다.
- 전송되는 데이터는 SSL/TLS 프로토콜을 사용하여 암호화합니다.

### 1.3 정기적인 보안 점검
- 보안 점검을 정기적으로 실시하여 취약점을 발견하고 수정합니다.
- 외부 보안 전문가에 의한 감사도 고려합니다.

## 2. 사용자 인증

### 2.1 인증 방법
- 기본 인증 방식으로는 이메일과 비밀번호 조합을 사용합니다.
- 추가적인 보안 강화를 위해 다중 인증(MFA)을 도입합니다. [

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 다음과 같습니다:

1. **보안 기본 원칙**
   - 최소 권한 원칙: 사용자는 자신의 업무에 필요한 최소한의 권한만 부여받아야 하며, 정기적으로 권한을 검토하고 불필요한 권한은 즉시 회수합니다.
   - 데이터 암호화: 모든 민감 데이터는 저장 시 암호화하여 보호하고, 전송되는 데이터는 SSL/TLS 프로토콜을 사용하여 암호화합니다.
   - 정기적인 보안 점검: 보안 점검을 정기적으로 실시하여 취약점을 발견하고 수정하며, 외부 보안 전문가에 의한 감사도 고려합니다.

2. **사용자 인증**
   - 기본 인증 방식으로는 이메일과 비밀번호 조합을 사용합니다.
   - 추가적인 보안 강화를 위해 다중 인증(MFA)을 도입합니다.

3. **로그 관리**
   - 모든 인증 시도 및 API 호출에 대한 로그를 기록하고, 로그에는 사용자 ID, IP 주소, 요청 시간, 요청 내용 등을 포함합니다.
   - 정기적으로 로그를 분석하여 비정상적인 활동을 탐지하고, 이상 징후 발견 시 즉시 대응합니다.

4. **사고 대응**
   - 보안 사고 발생 시 대응 절차를 문서화하고, 사고 발생 시 즉시 관련 팀에 통보하여 피해를 최소화하기 위한 조치를 취합니다.

5. **교육 및 훈련**
   - 모든 팀원은 보안 교육을 정기적으로 이수해야 하며, 보안 사고 대응 훈련을 통해 실제 상황에 대비합니다.

이 지침을 숙지하고 준수하여 안전한 시스템 운영에 기여해야 합니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: 

name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 에러 핸들링 매뉴얼
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 에러 핸들링의 중요성
에러 핸들링은 안정적인 시스템 운영을 위해 필수적입니다. 사용자 경험을 저해하지 않도록 적절한 에러 메시지를 제공하고, 시스템의 신뢰성을 높이는 것이 목표입니다.

## 2. 에러 종류
에러는 크게 두 가지로 분류됩니다.

### 2.1. 클라이언트 에러
- **정의**: 사용자의 요청에 문제가 있는 경우 발생
- **예시**: 잘못된 입력, 인증 실패 등

### 2.2. 서버 에러
- **정의**: 서버 내부에서 발생하는 문제
- **예시**: 데이터베이스 연결 실패, 외부 API 호출 실패 등

## 3. 에러 핸들링 단계
에러 핸들링을 위한 단계는 다음과 같습니다. [[ref1]]
---

다음 개정 제안: 에러 핸들링 매뉴얼에 실제 사례를 추가하여 이해도를 높일 수 있도록 제안합니다. 또한, 에러 로그 분석 결과를 공유하는 섹션을 추가하면 좋겠습니다. [[ref2]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 장애 대응 매뉴얼 (Runbook)
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 장애 발생 시 초기 대응 절차

1. **장애 인지**
   - 모니터링 시스템에서 장애 알림 수신
   - 사용자 불만 및 문의 확인

2. **문제 확인**
   - 장애 발생 서비스 및 증상 파악
   - 로그 및 모니터링 데이터 확인

3. **팀 알림**
   - 백엔드팀 전체에 장애 발생 사실 통보
   - 관련 팀(예: 프론트엔드, QA)에도 정보 공유

## 2. 장애 원인 분석

1. **로그 분석**
   - 장애 발생 시간대의 로그 확인
   - 에러 메시지 및 스택 트레이스 분석

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링의 중요성을 "안정적인 시스템 운영을 위해 필수적"이라고 설명하고 있습니다. 또한, 사용자 경험을 저해하지 않도록 적절한 에러 메시지를 제공하고 시스템의 신뢰성을 높이는 것이 목표라고 명시하고 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "v1.0 뭐에요?"}}']
INFO:utils2.service:cto search keyword: v1.0 뭐에요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 7. 개정 이력
- v1.0 — 2025-08-29: 초기 버전 작성

--- 

이 문서는 데이터/AI팀이 당장 실행 가능한 단계와 체크리스트를 포함하고 있습니다. 각 단계별로 필요한 조치를 취하여 모델 성능을 최적화하시기 바랍니다. [[ref1]]
## 7. 개정 이력
- v1.0 — 오늘

--- 

이 문서는 데이터/AI팀의 주간 업무 계획을 명확히 하고, 각 팀원이 실행 가능한 단계로 업무를 수행할 수 있도록 돕기 위해 작성되었습니다. 각 팀원은 자신의 작업을 주의 깊게 확인하고, 필요한 경우 상호 협력하여 목표를 달성하시기 바랍니다. [[ref2]]
## 6. 개정 이력
- v1.0 — 2025-08-29: 최초 작성

--- 

이 문서는 데이터/AI팀의 정기회의 내용을 바탕으로 작성되었습니다. 각 팀원은 Action Item에 따라 책임을 다하고, 정기적으로 진행 상황을 업데이트해 주시기 바랍니다. [[ref3]]


INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:v1.0은 문서의 초기 버전을 의미하며, 2025년 8월 29일에 작성되었습니다. 이 버전은 데이터/AI팀의 다양한 문서에서 사용되며, 각 문서의 개정 이력에 포함되어 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목들은 무엇인가요?"}}']
INFO:utils2.service:cto search keyword: 코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목들은 무엇인가요?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 체크리스트

- [ ] 코드 리뷰 완료
- [ ] 모든 테스트 통과
- [ ] 배포 환경 준비 완료
- [ ] 환경 변수 설정 완료
- [ ] 배포 실행 완료
- [ ] 모니터링 설정 완료
- [ ] 백업 진행
- [ ] 보안 점검 완료

이 가이드는 코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하는 데 도움을 줄 것입니다. 필요 시, 팀 내에서 지속적으로 업데이트하고 개선해 나가야 합니다.

---

다음 개정 제안: 배포 과정에서의 자동화 도구 사용에 대한 추가 정보를 포함할 수 있습니다. 또한, 보안 점검 방법을 좀 더 구체적으로 기술할 필요가 있습니다. [[ref1]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 배포/운영 가이드
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 목적
본 문서는 코드노바 백엔드팀의 배포 및 운영 과정을 명확히 하여 안정적이고 효율적인 서비스 제공을 목표로 합니다.

## 배포 준비 단계

1. **코드 검토**
   - 모든 변경 사항에 대해 코드 리뷰를 진행합니다.
   - 리뷰어는 최소 2명 이상 지정하여 품질을 보장합니다.

2. **테스트 수행**
   - 단위 테스트 및 통합 테스트를 실행합니다.
   - 테스트가 실패할 경우, 문제를 해결한 후 재테스트를 진행합니다.

3. **버전 관리**
   - 배포할 버전 번호를 확인합니다.
   - 변경 사항을 Changelog에 기록합니다.

## 배포 단계 [[ref2]]
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 백엔드팀 주간 업무 계획

**작성일**: 2025-08-29  
**작성자**: [작성자 이름]  
**팀**: 백엔드팀  

---

## 1. 주간 목표

- 서비스 안정성 강화
- 신규 기능 개발 및 배포
- 기술 부채 관리 및 리팩토링

## 2. 주요 업무

##

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목은 다음과 같습니다.

- 코드 리뷰 완료
- 모든 테스트 통과
- 배포 환경 준비 완료
- 환경 변수 설정 완료
- 배포 실행 완료
- 모니터링 설정 완료
- 백업 진행
- 보안 점검 완료

이 항목들은 배포 및 운영 과정에서 중요한 점검 사항으로, 필요 시 자동화 도구 사용이나 보안 점검 방법의 구체화도 고려할 수 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "챗봇 대화는 뭐 저장해?"}}']
INFO:utils2.service:cto search keyword: 챗봇 대화는 뭐 저장해?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
<!-- 회사: 코드노바 | 대상: 사원(백엔드) | 작성일: 2025-08-29 -->
# 데이터베이스 스키마 & 변경 이력 문서
분류: backend | 회사: 코드노바 | 버전: v1.0 | 작성일: 2025-08-29

## 1. 데이터베이스 스키마 개요

코드노바의 데이터베이스는 주요 서비스의 기능을 지원하기 위해 설계되었습니다. 데이터베이스는 다음과 같은 주요 엔티티로 구성됩니다:

- **사용자(User)**: 사용자 정보를 저장합니다.
- **글쓰기(Article)**: 생성된 글의 정보를 저장합니다.
- **이미지(Image)**: 업로드된 이미지 정보를 저장합니다.
- **요약(Summary)**: 생성된 요약 정보를 저장합니다.
- **챗봇 대화(Chat)**: 챗봇과의 대화 기록을 저장합니다.
- **광고(Ads)**: 대화형 광고 정보를 저장합니다.

각 엔티티는 서로 관계를 가지며, 데이터의 무결성을 유지하기 위해 외래 키 제약 조건을 설정합니다. [[ref1]]
### 2.4 요약(Summary)

- **summary_id**: 요약 고유 ID
- **article_id**: 관련 글 ID (글쓰기 테이블과 연결)
- **summary_text**: 요약 내용
- **created_at**: 생성 일자

### 2.5 챗봇 대화(Chat)

- **chat_id**: 대화 고유 ID
- **user_id**: 사용자 ID (사용자 테이블과 연결)
- **message**: 대화 내용
- **created_at**: 대화 일자

### 2.6 광고(Ads)

- **ad_id**: 광고 고유 ID
- **user_id**: 광고주 ID (사용자 테이블과 연결)
- **ad_content**: 광고 내용
- **created_at**: 생성 일자

## 3. 변경 이력

변경 이력은 데이터베이스 스키마의 변경 사항을 기록하여 추적할 수 있도록 합니다. 변경 사항이 있을 경우 아래와 같은 형식으로 기록합니다. [[

INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:챗봇 대화는 데이터베이스의 "챗봇 대화(Chat)" 테이블에 저장됩니다. 이 테이블에는 대화 고유 ID(chat_id), 사용자 ID(user_id), 대화 내용(message), 그리고 대화 일자(created_at)가 포함되어 있습니다.
INFO:utils2.service:Chat Request - Permission: cto, Tone: formal
INFO:utils2.service:Tools Prompt: 사용자는 cto로서, 모든 팀의 문서를 열람할 수 있는 개발팀 최고 관리자입니다. 질문이 들어오면 반드시 cto_search 툴을 호출하여 문서를 검색하세요.
INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:LLM Tool Parse Response Success
INFO:utils2.service:LLM Tools Match
INFO:utils2.service:['{"name": "cto_search", "arguments": {"keyword": "2025-08-01에 기록된 변경 사항은 무엇입니까?"}}']
INFO:utils2.service:cto search keyword: 2025-08-01에 기록된 변경 사항은 무엇입니까?


name='cto_search' description='모든 team(backend, frontend, data_ai)의 문서를 검색합니다.' args_schema=<class 'langchain_core.utils.pydantic.cto_search'> func=<function cto_search at 0x00000237DC2BA3E0>


INFO:utils2.service:cto search found 3 results.
INFO:utils2.service:cto_search tool Response Success


검색 결과:
-----
## 6. 개정 이력
- v1.0 — 2025-08-29: 최초 작성

--- 

이 문서는 데이터/AI팀의 정기회의 내용을 바탕으로 작성되었습니다. 각 팀원은 Action Item에 따라 책임을 다하고, 정기적으로 진행 상황을 업데이트해 주시기 바랍니다. [[ref1]]
### 3.1 변경 이력 기록 양식

- **변경 날짜**: YYYY-MM-DD
- **변경 사항**: 구체적인 변경 내용
- **작성자**: 변경을 수행한 개발자 이름

### 3.2 예시

- **변경 날짜**: 2025-08-01
- **변경 사항**: 사용자 테이블에 'last_login' 필드 추가
- **작성자**: 홍길동

## 4. 검증 포인트

데이터베이스 스키마 변경 후 다음 사항을 검증합니다:

- 모든 테이블과 필드가 올바르게 생성되었는지 확인
- 외래 키 제약 조건이 제대로 설정되었는지 확인
- 기존 데이터와의 호환성 검증
- 쿼리 성능 테스트 수행

## 5. 결론

본 문서는 코드노바의 데이터베이스 스키마 및 변경 이력에 대한 기본 정보를 제공합니다. 향후 변경 사항이 있을 경우, 반드시 변경 이력을 기록하고 검증 포인트를 확인하여 데이터의 무결성을 유지해야 합니다.

--- [[ref2]]
## 7. 개정 이력
- v1.0 — 2025-08-29: 초기 버전 작성

--- 

이 문서는 데이터/AI팀이 당장 실행 가능한 단계와 체크리스트를 포함하고 있습니다. 각 단계별로 필요한 조치를 취하여 모델 성능을 최적화하시기 바랍니다. [[ref3]]


INFO:httpx:HTTP Request: POST https://qfs7408cldsqap-8000.proxy.runpod.net/v1/chat/completions "HTTP/1.1 200 OK"
INFO:utils2.service:Final Assistant Reply Generated
INFO:utils2.service:2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필드가 추가된 것입니다. 해당 변경 사항은 홍길동이 작성한 것입니다.


In [15]:
display(ragas_evaluated_dataset.to_pandas())

user_input  \
0                          크랙(Crack) 앱의 주요 기능은 무엇인가요?   
1                                RabbitMQ의 역할은 무엇인가요?   
2                           프론트엔드 팀에서 JWT의 역할은 무엇인가요?   
3         코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?   
4                                         MFA는 뭐하는거야?   
5                           JWT는 API 접근 시 어떤 역할을 합니까?   
6       코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서 에러 핸들링의 중...   
8                                           v1.0 뭐에요?   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10                                      챗봇 대화는 뭐 저장해?   
11                      2025-08-01에 기록된 변경 사항은 무엇입니까?   

                                   retrieved_contexts  \
0   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
1   [검색 결과:\n-----\n### 2.2. 데이터베이스\n- **역할**: 사용자...   
2   [검색 결과:\n-----\n## 6. 개정 이력\n- v1.0 — 오늘\n\n--...   
3   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
4   [검색 결과:\n-----\n## 5. 리스크 한계 및 보완책\n- **리스크 한계...   
5   [검색 결과:\n-----\n## 3. 인증 및 권한 관리\n- **인증 방식**:...   
6   [검색 결과:\n-----\n## 4. 로그 관리\n\n### 4.1 로그 기록\n...   
7   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
8   [검색 결과:\n-----\n## 7. 개정 이력\n- v1.0 — 2025-08-...   
9   [검색 결과:\n-----\n## 체크리스트\n\n- [ ] 코드 리뷰 완료\n- ...   
10  [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
11  [검색 결과:\n-----\n## 6. 개정 이력\n- v1.0 — 2025-08-...   

                                             response  \
0   크랙(Crack) 앱의 주요 기능은 생성형 AI를 활용한 글쓰기, 이미지 생성, 요...   
1   RabbitMQ는 메시지 큐로, 비동기 작업 처리를 위해 시스템 간 메시지를 전달하...   
2   검색 결과에 프론트엔드 팀에서 JWT의 역할에 대한 구체적인 내용은 포함되어 있지 ...   
3   코드노바의 백엔드 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI ...   
4   MFA는 다중 인증(Multi-Factor Authentication)의 약자로, ...   
5   JWT(JSON Web Token)는 API 접근 시 사용자의 인증을 처리하는 데 ...   
6   코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 다음과 같습니...   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링의 ...   
8   v1.0은 문서의 초기 버전을 의미하며, 2025년 8월 29일에 작성되었습니다. ...   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10  챗봇 대화는 데이터베이스의 "챗봇 대화(Chat)" 테이블에 저장됩니다. 이 테이블...   
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...   

                                            reference  
0   크랙(Crack)은 AI 페르소나 챗봇 앱으로, 코드노바의 서비스 아키텍처에서 생성...  
1   RabbitMQ는 비동기 작업 처리를 위해 시스템 간 메시지를 전달하는 메시지 큐입니다.  
2          JWT(JSON Web Token)는 사용자 인증을 처리하는 데 사용됩니다.  
3   코드노바의 백엔드 서비스 아키텍처는 확장 가능하고 안정적인 서비스를 제공하기 위해 ...  
4               MFA는 추가적인 보안 강화를 위해 도입되는 다중 인증 방법입니다.  
5   JWT(JSON Web Token)는 API 접근 시 인증을 위해 사용되며, 토큰은...  
6   이 가이드는 코드노바의 백엔드 개발팀이 보안 및 인증을 효과적으로 관리하기 위한 기...  
7   코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링이 안정적인 시스템 운영을 위해 필수...  
8   v1.0은 코드노바 백엔드팀의 배포 및 운영 가이드의 버전 번호로, 이 문서는 안정...  
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...  
10                          챗봇 대화는 챗봇과의 대화 기록을 저장합니다.  
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...

### RAGAS 평가 실행

**ragas.metrics 주요 클래스 설명**

- **LLMContextRecall**  
  
  주어진 답변이 원문(문맥)에서 제시된 정보를 얼마나 잘 회상(recall)했는지 평가하는 클래스이다.
  
  예를 들어, 원문에 중요한 사실 10개가 있는데 답변이 그중 몇 개를 잘 포함했는지 측정한다.
  즉, 잊지 않고 잘 회상했는지 보는 지표다.

- **Faithfulness**  
  답변이 원본(문맥) 정보에 기반하여 얼마나 충실하고 일관되게 생성되었는지 평가하는 클래스이다.

  답변 내 내용이 문맥과 어긋나지 않고, 문맥에서 유추 가능한 사실들로만 이루어졌는지를 본다.

  낮은 faithfulness 값은 모델이 문맥에 없는 정보를 만들어내거나 왜곡하여 답변함을 의미할 수 있어, 환각 발생 정도를 간접적으로 보여준다.

- **FactualCorrectness**  
  답변의 내용이 실제 사실과 얼마나 정확하게 일치하는지를 평가하는 클래스이다.

  문맥뿐 아니라 절대적인 사실관계(정확성)와 맞는지 따진다.

  즉, 문맥이 아니라 객관적인 사실에 맞는지를 측정하는 지표다.

In [13]:
# LLM 래퍼 생성
evaluator_llm = ChatOpenAI(model="gpt-4.1-mini", temperature=0)
evaluator_llm = LangchainLLMWrapper(evaluator_llm)

# 평가 메트릭 정의
metrics = [
    LLMContextRecall(),    # 검색된 컨텍스트의 회수율
    Faithfulness(),        # 생성된 답변의 충실도
    FactualCorrectness()   # 사실적 정확성
]

In [14]:
# 평가 실행
result = evaluate(
    dataset=ragas_evaluated_dataset,  # 평가 데이터셋
    metrics=metrics,             # 평가 메트릭
    llm=evaluator_llm,          # LLM 래퍼
)

print(result)

Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"

{'context_recall': 0.8333, 'faithfulness': 0.7688, 'factual_correctness(mode=f1)': 0.4425}


In [17]:
# 결과를 DataFrame으로 변환
result_df = result.to_pandas()

# 결과 저장
result_df.to_csv('ragas_evaluation_result.csv', index=False)

display(result_df)

user_input  \
0                          크랙(Crack) 앱의 주요 기능은 무엇인가요?   
1                                RabbitMQ의 역할은 무엇인가요?   
2                           프론트엔드 팀에서 JWT의 역할은 무엇인가요?   
3         코드노바의 백엔드 서비스 아키텍처는 어떻게 설계되었고, 그 목적은 무엇인가요?   
4                                         MFA는 뭐하는거야?   
5                           JWT는 API 접근 시 어떤 역할을 합니까?   
6       코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 무엇입니까?   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서 에러 핸들링의 중...   
8                                           v1.0 뭐에요?   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10                                      챗봇 대화는 뭐 저장해?   
11                      2025-08-01에 기록된 변경 사항은 무엇입니까?   

                                   retrieved_contexts  \
0   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
1   [검색 결과:\n-----\n### 2.2. 데이터베이스\n- **역할**: 사용자...   
2   [검색 결과:\n-----\n## 6. 개정 이력\n- v1.0 — 오늘\n\n--...   
3   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
4   [검색 결과:\n-----\n## 5. 리스크 한계 및 보완책\n- **리스크 한계...   
5   [검색 결과:\n-----\n## 3. 인증 및 권한 관리\n- **인증 방식**:...   
6   [검색 결과:\n-----\n## 4. 로그 관리\n\n### 4.1 로그 기록\n...   
7   [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
8   [검색 결과:\n-----\n## 7. 개정 이력\n- v1.0 — 2025-08-...   
9   [검색 결과:\n-----\n## 체크리스트\n\n- [ ] 코드 리뷰 완료\n- ...   
10  [검색 결과:\n-----\n<!-- 회사: 코드노바 | 대상: 사원(백엔드) | ...   
11  [검색 결과:\n-----\n## 6. 개정 이력\n- v1.0 — 2025-08-...   

                                             response  \
0   크랙(Crack) 앱의 주요 기능은 생성형 AI를 활용한 글쓰기, 이미지 생성, 요...   
1   RabbitMQ는 메시지 큐로, 비동기 작업 처리를 위해 시스템 간 메시지를 전달하...   
2   검색 결과에 프론트엔드 팀에서 JWT의 역할에 대한 구체적인 내용은 포함되어 있지 ...   
3   코드노바의 백엔드 서비스 아키텍처는 생성형 AI 글쓰기·이미지·요약 플랫폼, AI ...   
4   MFA는 다중 인증(Multi-Factor Authentication)의 약자로, ...   
5   JWT(JSON Web Token)는 API 접근 시 사용자의 인증을 처리하는 데 ...   
6   코드노바의 백엔드 개발팀이 보안 및 인증을 관리하기 위한 기본 지침은 다음과 같습니...   
7   2025년 8월 29일에 작성된 코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링의 ...   
8   v1.0은 문서의 초기 버전을 의미하며, 2025년 8월 29일에 작성되었습니다. ...   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...   
10  챗봇 대화는 데이터베이스의 "챗봇 대화(Chat)" 테이블에 저장됩니다. 이 테이블...   
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...   

                                            reference  context_recall  \
0   크랙(Crack)은 AI 페르소나 챗봇 앱으로, 코드노바의 서비스 아키텍처에서 생성...             1.0   
1   RabbitMQ는 비동기 작업 처리를 위해 시스템 간 메시지를 전달하는 메시지 큐입니다.             1.0   
2          JWT(JSON Web Token)는 사용자 인증을 처리하는 데 사용됩니다.             0.0   
3   코드노바의 백엔드 서비스 아키텍처는 확장 가능하고 안정적인 서비스를 제공하기 위해 ...             1.0   
4               MFA는 추가적인 보안 강화를 위해 도입되는 다중 인증 방법입니다.             1.0   
5   JWT(JSON Web Token)는 API 접근 시 인증을 위해 사용되며, 토큰은...             1.0   
6   이 가이드는 코드노바의 백엔드 개발팀이 보안 및 인증을 효과적으로 관리하기 위한 기...             1.0   
7   코드노바의 에러 핸들링 매뉴얼에서는 에러 핸들링이 안정적인 시스템 운영을 위해 필수...             1.0   
8   v1.0은 코드노바 백엔드팀의 배포 및 운영 가이드의 버전 번호로, 이 문서는 안정...             0.0   
9   코드노바 백엔드팀이 배포 및 운영을 효율적으로 수행하기 위해 필요한 체크리스트 항목...             1.0   
10                          챗봇 대화는 챗봇과의 대화 기록을 저장합니다.             1.0   
11  2025-08-01에 기록된 변경 사항은 사용자 테이블에 'last_login' 필...             1.0   

    faithfulness  factual_correctness(mode=f1)  
0       0.333333                          0.25  
1       0.400000                          0.67  
2       0.500000                          0.50  
3       0.777778                          0.12  
4       0.750000                          0.67  
5       0.714286                          0.40  
6       1.000000                          0.17  
7       1.000000                          0.67  
8       0.750000                          0.00  
9       1.000000                          0.86  
10      1.000000                          0.00  
11      1.000000                          1.00

### 평가지표 해석 및 개선방안

1. context_recall (문맥 리콜, 0~1):

    검색된 문서나 문맥에 답변 작성을 위한 필요한 정보가 얼마나 포함되어 있는지를 나타냅니다. 값이 0.5라면, 필요한 정보가 절반 정도는 검색되어 있다는 뜻으로, 문맥의 충분성이 보통 수준임을 뜻합니다.

2. faithfulness (충실도, 0~1):

    생성된 답변이 검색된 문맥 내 정보에 얼마나 사실적으로 충실한지를 평가합니다. 0.5 정도면 절반 정도 문맥 정보를 충실히 반영했으나, 환각(허위 정보) 발생 가능성이 중간 정도임을 의미합니다.

3. factual_correctness (사실 정확성, F1 점수, 0~1)

    생성 답변의 사실적 정확도를 정밀도와 재현율의 조합으로 평가하는 지표입니다. 0.17은 비교적 낮은 점수로, 생성 답변에 사실과 다른 정보가 많이 포함되었을 가능성이 높음을 뜻합니다. 즉, 환각 현상이 꽤 존재할 수 있습니다.


**개선 방안**

- **검색 성능 향상**: 더 정확한 임베딩 모델이나 하이브리드 검색 방법을 고려한다
- **프롬프트 최적화**: 더 구체적이고 명확한 지시사항을 포함한 프롬프트를 설계한다
- **데이터 품질 개선**: 더 정확하고 완전한 참조 데이터를 확보한다
- **후처리 단계 추가**: 사실 검증 단계를 추가하여 정확성을 향상시킨다
- **다양한 검색 전략**(하이브리드 검색, 리랭킹 등) 실험
- **커스텀 평가 메트릭 개발**
- **실시간 평가 시스템 구축**
- **A/B 테스트를 통한 시스템 최적화**